File: frontend/src/utils/hoursParser.ts
/**
 * Parse the multi-line `hours_summary` string scraped from mrt.com.my into
 * structured fields the UI can render compactly.
 *
 * Example raw input:
 *   "Station open: 06:00 am
 *    Station closed: 12:00 am (Mon - Sat) / 11:25 pm (Sunday & PH)
 *    Last Train to Gombak: 12:12 am (Mon - Sat) / 11:42 pm (Sunday & PH)
 *    Last Train to Putra Height: 12:07 am (Mon - Sat) / 11:49 pm (Sunday & PH)"
 */

export interface ConditionalTime {
  time: string;
  condition: string;
}

export interface ParsedHours {
  open: ConditionalTime[];
  close: ConditionalTime[];
  lastTrains: { to: string; values: ConditionalTime[] }[];
  other: string[];
}

export function parseHoursSummary(raw: string | null | undefined): ParsedHours | null {
  if (!raw) return null;

  const result: ParsedHours = {
    open: [],
    close: [],
    lastTrains: [],
    other: [],
  };
  const lines = raw
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter(Boolean);
  if (!lines.length) return null;

  for (const line of lines) {
    const colon = line.indexOf(':');
    if (colon < 0) {
      result.other.push(line);
      continue;
    }
    const key = line.slice(0, colon).trim();
    const value = line.slice(colon + 1).trim();

    if (/^station open$/i.test(key)) {
      result.open.push(...splitConditionTime(value));
    } else if (/^station closed$/i.test(key)) {
      result.close.push(...splitConditionTime(value));
    } else if (/^last train to (.+)$/i.test(key)) {
      const m = key.match(/^last train to\s*(.+)$/i);
      const dest = m ? m[1].trim() : key;
      result.lastTrains.push({ to: dest, values: splitConditionTime(value) });
    } else if (/^last train$/i.test(key)) {
      // Some stations (e.g. monorail) show a single "Last train" line with no
      // explicit destination. Surface as an "other" note to avoid misplacing it.
      result.other.push(`${key}: ${value}`);
    } else {
      result.other.push(line);
    }
  }

  if (
    !result.open.length &&
    !result.close.length &&
    !result.lastTrains.length &&
    !result.other.length
  ) {
    return null;
  }
  return result;
}

/**
 * Split values such as "12:00 am (Mon - Sat) / 11:25 pm (Sunday & PH)" into
 * an array of {time, condition} pairs. Falls back to a single entry with
 * empty condition when no parentheses are present.
 */
function splitConditionTime(raw: string): ConditionalTime[] {
  return raw
    .split(/\s*\/\s*/)
    .map((segment) => {
      const trimmed = segment.trim();
      const m = trimmed.match(/^(.+?)\s*\((.+?)\)\s*$/);
      if (m) {
        return { time: normalizeTime(m[1].trim()), condition: normalizeCondition(m[2].trim()) };
      }
      return { time: normalizeTime(trimmed), condition: '' };
    })
    .filter((entry) => entry.time);
}

/**
 * Light normalization: uppercase AM/PM, collapse whitespace.
 * We deliberately avoid converting 12:00 am vs 12:00 midnight semantics —
 * preserving the source string keeps trust in shown values.
 */
function normalizeTime(raw: string): string {
  return raw
    .replace(/\s+/g, ' ')
    .replace(/\b(am|pm)\b/gi, (m) => m.toUpperCase())
    .trim();
}

function normalizeCondition(raw: string): string {
  return raw
    .replace(/\s+/g, ' ')
    .replace(/\bSunday\b/gi, 'Sun')
    .replace(/\bMonday\b/gi, 'Mon')
    .replace(/\b& PH\b/gi, '& PH')
    .trim();
}

File: backend/app/utils/hours_parser.py
"""Parse MRT-style station hours_summary strings into structured sections."""

from __future__ import annotations

import re
from dataclasses import dataclass, field


@dataclass
class ConditionalTime:
    time: str
    condition: str = ""


@dataclass
class LastTrainDest:
    to: str
    values: list[ConditionalTime] = field(default_factory=list)


@dataclass
class ParsedHours:
    open: list[ConditionalTime] = field(default_factory=list)
    close: list[ConditionalTime] = field(default_factory=list)
    last_trains: list[LastTrainDest] = field(default_factory=list)
    other: list[str] = field(default_factory=list)


def parse_hours_summary(raw: str | None) -> ParsedHours | None:
    if not raw or not raw.strip():
        return None

    result = ParsedHours()
    lines = [line.strip() for line in raw.replace("\r\n", "\n").split("\n") if line.strip()]
    if not lines:
        return None

    for line in lines:
        colon = line.find(":")
        if colon < 0:
            result.other.append(line)
            continue
        key = line[:colon].strip()
        value = line[colon + 1 :].strip()

        if re.match(r"^station open$", key, re.I):
            result.open.extend(_split_condition_time(value))
        elif re.match(r"^station closed$", key, re.I):
            result.close.extend(_split_condition_time(value))
        elif m := re.match(r"^last train to\s*(.+)$", key, re.I):
            dest = m.group(1).strip()
            result.last_trains.append(LastTrainDest(to=dest, values=_split_condition_time(value)))
        elif re.match(r"^last train$", key, re.I):
            result.other.append(f"{key}: {value}")
        else:
            result.other.append(line)

    if not (result.open or result.close or result.last_trains or result.other):
        return None
    return result


def _split_condition_time(raw: str) -> list[ConditionalTime]:
    entries: list[ConditionalTime] = []
    for segment in re.split(r"\s*/\s*", raw):
        trimmed = segment.strip()
        if not trimmed:
            continue
        m = re.match(r"^(.+?)\s*\((.+?)\)\s*$", trimmed)
        if m:
            entries.append(
                ConditionalTime(
                    time=_normalize_time(m.group(1).strip()),
                    condition=_normalize_condition(m.group(2).strip()),
                )
            )
        else:
            entries.append(ConditionalTime(time=_normalize_time(trimmed)))
    return [e for e in entries if e.time]


def _normalize_time(raw: str) -> str:
    return re.sub(r"\b(am|pm)\b", lambda m: m.group(1).upper(), re.sub(r"\s+", " ", raw)).strip()


def _normalize_condition(raw: str) -> str:
    cleaned = re.sub(r"\s+", " ", raw).strip()
    cleaned = re.sub(r"\bSunday\b", "Sun", cleaned, flags=re.I)
    cleaned = re.sub(r"\bMonday\b", "Mon", cleaned, flags=re.I)
    return cleaned
